In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import *
import math

In [ ]:
# create a model
model = ConcreteModel()

In [ ]:
# declare decision variables with initialize=2.
model.x1 = Var(initialize=2.)
model.x2 = Var(initialize=2.)
model.x3 = Var(initialize=2.)
model.x4 = Var(initialize=2.)
model.x5 = Var(initialize=2.)

model.constr1 = Constraint(expr=(model.x1)+model.x2**2+model.x3**3-2-3*math.sqrt(2)== 0)
model.constr2 = Constraint(expr=model.x2-model.x3**2+model.x4+2-2*math.sqrt(2) == 0)
model.constr3 = Constraint(expr=model.x1*model.x5-2 == 0)
model.x1.setlb(-5)
model.x1.setub(5)
model.x3.setlb(0)
model.x3.setub(3)

In [ ]:
def func(model):
  return (model.x1-1)**2 + (model.x1 - model.x2)**2 + (model.x2 - model.x3)**2 + (model.x3 - model.x4)**4 + (model.x4 - model.x5)**4

In [ ]:
# declare objective
model.cost = Objective(rule = func, sense=minimize)

In [ ]:
model.pprint()

5 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :    -5 :   2.0 :     5 : False : False :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :   2.0 :  None : False : False :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :   2.0 :     3 : False : False :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :   2.0 :  None : False : False :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :   2.0 :  None : False : False :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : (x1 - 1)**2 + (x1 - x2)**2 + (x2 - x3)**2 + (x3 - x4)**4 + (x4 - x5)**4

In [ ]:
#We will use ipopt solver for solving the optimization problem

!wget -N -q "https://ampl.com/dl/open/ipopt/ipopt-linux64.zip"
!unzip -o -q ipopt-linux64


In [ ]:
solver = SolverFactory('ipopt', executable='/content/ipopt')

In [ ]:
results = solver.solve(model)

In [ ]:
#display the statistics
print(results)


Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 5
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.12.13\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.026993751525878906
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [ ]:
# display solution
print('\nCost = ', model.cost())

print('\nOptimal solution:')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())


print('\nConstraints')
print('Constraint 1  = ', model.constr1())
print('Constraint 2  = ', model.constr2())
print('Constraint 3  = ', model.constr3())


Cost =  0.07877682087105699

Optimal solution:
x1 =  1.1911274562403504
x2 =  1.3626031649434767
x3 =  1.4728179315306016
x4 =  1.6350166192407931
x5 =  1.6790814362660684

Constraints
Constraint 1  =  1.7763568394002505e-15
Constraint 2  =  -4.440892098500626e-16
Constraint 3  =  -3.774758283725532e-15


In [ ]:
#change the initial value
model.x1 = Var(initialize=4.)
model.x2 = Var(initialize=4.)
model.x3 = Var(initialize=4.)
model.x4 = Var(initialize=4.)
model.x5 = Var(initialize=4.)

model.constr1 = Constraint(expr=(model.x1)+model.x2**2+model.x3**3-2-3*math.sqrt(2)== 0)
model.constr2 = Constraint(expr=model.x2-model.x3**2+model.x4+2-2*math.sqrt(2) == 0)
model.constr3 = Constraint(expr=model.x1*model.x5-2 == 0)
model.x1.setlb(-5)
model.x1.setub(5)
model.x3.setlb(0)
model.x3.setub(3)

In [ ]:
# declare objective
model.cost = Objective(rule = func, sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
results = solver.solve(model)
print(results)


Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 5
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.12.13\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.025893211364746094
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [ ]:
#display solution
print('\nCost = ', model.cost())

print('\nOptimal solution:')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())


print('\nConstraints')
print('Constraint 1  = ', model.constr1())
print('Constraint 2  = ', model.constr2())
print('Constraint 3  = ', model.constr3())


Cost =  0.07877682087105696

Optimal solution:
x1 =  1.1911274562403502
x2 =  1.3626031649434764
x3 =  1.4728179315306016
x4 =  1.6350166192407936
x5 =  1.6790814362660686

Constraints
Constraint 1  =  8.881784197001252e-16
Constraint 2  =  -4.440892098500626e-16
Constraint 3  =  -3.9968028886505635e-15


In [ ]:
#change the initial value
model.x1 = Var(initialize=2**4.)
model.x2 = Var(initialize=2**4.)
model.x3 = Var(initialize=2**4.)
model.x4 = Var(initialize=2**4.)
model.x5 = Var(initialize=2**4.)

model.constr1 = Constraint(expr=(model.x1)+model.x2**2+model.x3**3-2-3*math.sqrt(2)== 0)
model.constr2 = Constraint(expr=model.x2-model.x3**2+model.x4+2-2*math.sqrt(2) == 0)
model.constr3 = Constraint(expr=model.x1*model.x5-2 == 0)
model.x1.setlb(-5)
model.x1.setub(5)
model.x3.setlb(0)
model.x3.setub(3)

In [ ]:
# declare objective
model.cost = Objective(rule = func, sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
results = solver.solve(model)
print(results)


Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 5
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.12.13\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.03658628463745117
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [ ]:
#display solution
print('\nCost = ', model.cost())

print('\nOptimal solution:')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())


print('\nConstraints')
print('Constraint 1  = ', model.constr1())
print('Constraint 2  = ', model.constr2())
print('Constraint 3  = ', model.constr3())


Cost =  0.07877682087105699

Optimal solution:
x1 =  1.1911274562403504
x2 =  1.3626031649434767
x3 =  1.4728179315306016
x4 =  1.6350166192407931
x5 =  1.6790814362660684

Constraints
Constraint 1  =  1.7763568394002505e-15
Constraint 2  =  -4.440892098500626e-16
Constraint 3  =  -3.774758283725532e-15


In [ ]:
#change the initial value
model.x1 = Var(initialize=2**8.)
model.x2 = Var(initialize=2**8.)
model.x3 = Var(initialize=2**8.)
model.x4 = Var(initialize=2**8.)
model.x5 = Var(initialize=2**8.)

model.constr1 = Constraint(expr=(model.x1)+model.x2**2+model.x3**3-2-3*math.sqrt(2)== 0)
model.constr2 = Constraint(expr=model.x2-model.x3**2+model.x4+2-2*math.sqrt(2) == 0)
model.constr3 = Constraint(expr=model.x1*model.x5-2 == 0)
model.x1.setlb(-5)
model.x1.setub(5)
model.x3.setlb(0)
model.x3.setub(3)

In [ ]:
# declare objective
model.cost = Objective(rule = func, sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
results = solver.solve(model)
print(results)


Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 5
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.12.13\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.06071352958679199
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [ ]:
#display solution
print('\nCost = ', model.cost())

print('\nOptimal solution:')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())


print('\nConstraints')
print('Constraint 1  = ', model.constr1())
print('Constraint 2  = ', model.constr2())
print('Constraint 3  = ', model.constr3())


Cost =  0.07877682087142031

Optimal solution:
x1 =  1.1911274559458875
x2 =  1.3626031648627808
x3 =  1.4728179316111962
x4 =  1.635016619557219
x5 =  1.6790814366787805

Constraints
Constraint 1  =  1.0102141345669224e-11
Constraint 2  =  -1.6728840535051859e-12
Constraint 3  =  -2.8383961847566752e-12


In [ ]:
#change the initial value
model.x1 = Var(initialize=2**10.)
model.x2 = Var(initialize=2**10.)
model.x3 = Var(initialize=2**10.)
model.x4 = Var(initialize=2**10.)
model.x5 = Var(initialize=2**10.)

model.constr1 = Constraint(expr=(model.x1)+model.x2**2+model.x3**3-2-3*math.sqrt(2)== 0)
model.constr2 = Constraint(expr=model.x2-model.x3**2+model.x4+2-2*math.sqrt(2) == 0)
model.constr3 = Constraint(expr=model.x1*model.x5-2 == 0)
model.x1.setlb(-5)
model.x1.setub(5)
model.x3.setlb(0)
model.x3.setub(3)

In [ ]:
# declare objective
model.cost = Objective(rule = func, sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
results = solver.solve(model)
print(results)


Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 5
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.12.13\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.0800323486328125
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [ ]:
#display solution
print('\nCost = ', model.cost())

print('\nOptimal solution:')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())


print('\nConstraints')
print('Constraint 1  = ', model.constr1())
print('Constraint 2  = ', model.constr2())
print('Constraint 3  = ', model.constr3())


Cost =  0.0787768208710788

Optimal solution:
x1 =  1.1911274548626596
x2 =  1.3626031645866354
x3 =  1.4728179318918395
x4 =  1.6350166206615917
x5 =  1.6790814382064987

Constraints
Constraint 1  =  6.288303211476887e-13
Constraint 2  =  -1.1857181902996672e-13
Constraint 3  =  -1.9590995492535512e-12


In [ ]:
#change the initial value
model.x1 = Var(initialize=2**12.)
model.x2 = Var(initialize=2**12.)
model.x3 = Var(initialize=2**12.)
model.x4 = Var(initialize=2**12.)
model.x5 = Var(initialize=2**12.)

model.constr1 = Constraint(expr=(model.x1)+model.x2**2+model.x3**3-2-3*math.sqrt(2)== 0)
model.constr2 = Constraint(expr=model.x2-model.x3**2+model.x4+2-2*math.sqrt(2) == 0)
model.constr3 = Constraint(expr=model.x1*model.x5-2 == 0)
model.x1.setlb(-5)
model.x1.setub(5)
model.x3.setlb(0)
model.x3.setub(3)

In [ ]:
# declare objective
model.cost = Objective(rule = func, sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
results = solver.solve(model)
print(results)


Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 5
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.12.13\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.0418093204498291
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [ ]:
#display solution
print('\nCost = ', model.cost())

print('\nOptimal solution:')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())


print('\nConstraints')
print('Constraint 1  = ', model.constr1())
print('Constraint 2  = ', model.constr2())
print('Constraint 3  = ', model.constr3())


Cost =  0.07877682153434043

Optimal solution:
x1 =  1.1911274444329922
x2 =  1.362603152254644
x3 =  1.4728179414916138
x4 =  1.635016658269562
x5 =  1.6790814465461419

Constraints
Constraint 1  =  1.8435020621154763e-08
Constraint 2  =  -3.00157898536213e-09
Constraint 3  =  -7.58064211403564e-09


**Report on Exercise 1 part 3:**

Solver status: ok

Time taken to solve: 0.026993751525878906
  
Solver termination condition: Optimal

**Report on Exercise 1 part 4:**

Cost =  0.07877682087105699

Optimal solution:

x1 =  1.1911274562403504

x2 =  1.3626031649434767

x3 =  1.4728179315306016

x4 =  1.6350166192407931

x5 =  1.6790814362660684

**Report on Exercise 1 part 5:**

Constraint 1, $x_1 + x_2^2 + x_3^3 - 2 - 4.242640687119286$ evaluates to
1.7763568394002505e-15 which is almost equal to zero = rhs, hence active.

Constraint 2, $x_2 - x_3^2 + x_4 + 2 - 2.8284271247461903 $ evaluates to
-4.440892098500626e-16 which is almost equal to zero = rhs, hence active.

Constraint 3, $x_1*x_5 - 2 $ evaluates to -3.774758283725532e-15 which is almost equal to zero = rhs, hence active.

I have taken the bound constraints as bound not as the constraints.

**Report on Exercise 1 part 6:**

**When i = 2**,

**Cost** =  0.07877682087105696

**Optimal solution:**

x1 =  1.1911274562403502

x2 =  1.3626031649434764

x3 =  1.4728179315306016

x4 =  1.6350166192407936

x5 =  1.6790814362660686

**Time taken to solve:** 0.025893211364746094

**When i = 4**,

**Cost** =  0.07877682087105699

**Optimal solution:**

x1 =  1.1911274562403504

x2 =  1.3626031649434767

x3 =  1.4728179315306016

x4 =  1.6350166192407931

x5 =  1.6790814362660684

**Time taken to solve:** 0.03658628463745117

**When i = 8**,

**Cost** =  0.07877682087142031

**Optimal solution:**

x1 =  1.1911274559458875

x2 =  1.3626031648627808

x3 =  1.4728179316111962

x4 =  1.635016619557219

x5 =  1.6790814366787805

**Time taken to solve:** 0.06071352958679199

**When i = 10**,

**Cost** =  0.0787768208710788

**Optimal solution:**

x1 =  1.1911274548626596

x2 =  1.3626031645866354

x3 =  1.4728179318918395

x4 =  1.6350166206615917

x5 =  1.6790814382064987

**Time taken to solve:** 0.0800323486328125

**When i = 12**,

**Cost** =  0.07877682153434043

**Optimal solution:**

x1 =  1.1911274444329922

x2 =  1.362603152254644

x3 =  1.4728179414916138

x4 =  1.635016658269562

x5 =  1.6790814465461419

**Time taken to solve:** 0.0418093204498291

**Observations:**

1. The objective function value in all the cases is almost the same (only last few digits changed)

2. The value of all the constraints in all the cases is also the same (only last few digits changed)

3. For i = 2 to i = 10, the time taken by the solver to solve the problem increased gradually but for i = 12, it increased but time taken was greater than time taken in case of i = 2 and i = 4